homeassistant/components/media_player/sonos.py

```diff
@@ -141,46 +139,42 @@ def safe_parse_event_xml(xml):
     if DATA_SONOS not in hass.data:
         hass.data[DATA_SONOS] = SonosData()
 
-    advertise_addr = config.get(CONF_ADVERTISE_ADDR, None)
-    if advertise_addr:
-        soco.config.EVENT_ADVERTISE_IP = advertise_addr
-
+    players = []
     if discovery_info:
         player = soco.SoCo(discovery_info.get('host'))
 
         # If device already exists by config
         if player.uid in hass.data[DATA_SONOS].uids:
             return
 
-        if player.is_visible:
-            hass.data[DATA_SONOS].uids.add(player.uid)
-            add_devices([SonosDevice(player)])
+        # If invisible, such as a stereo slave
+        if not player.is_visible:
+            return
+
+        players.append(player)
     else:
-        players = None
-        hosts = config.get(CONF_HOSTS, None)
+        hosts = config.get(CONF_HOSTS)
         if hosts:
             # Support retro compatibility with comma separated list of hosts
             # from config
             hosts = hosts[0] if len(hosts) == 1 else hosts
             hosts = hosts.split(',') if isinstance(hosts, str) else hosts
-            players = []
             for host in hosts:
                 try:
                     players.append(soco.SoCo(socket.gethostbyname(host)))
                 except OSError:
                     _LOGGER.warning("Failed to initialize '%s'", host)
-
-        if not players:
+        else:
             players = soco.discover(
                 interface_addr=config.get(CONF_INTERFACE_ADDR))
 
         if not players:
             _LOGGER.warning("No Sonos speakers found")
             return
 
-        hass.data[DATA_SONOS].uids.update([p.uid for p in players])
-        add_devices([SonosDevice(p) for p in players])
-        _LOGGER.debug("Added %s Sonos speakers", len(players))
+    hass.data[DATA_SONOS].uids.update([p.uid for p in players])
```

Comment:
No need to add the square brackets. `hass.data[DATA_SONOS].uids.update(p.uid for p in players)` will pass it as a generator which works just fine.